File: TPII-2025-2/PadraoGoF/Estrutural/Adapter/Adapter01.py
# 1 - Interface do Cliente
class AplicacaoCliente:
    def request(self):
        pass

class Cliente:
    def __init__(self, target: AplicacaoCliente):
        self.target = target

    def cria_request(self):
        print("Cliente - Fazendo uma requisição")
        self.target.request()


# 2 - Serviço Existente
class Adaptee:
    def especifica_request(self):
        print("Requisição específica do Adaptee")


# 3 - Adapter
class Adapter(AplicacaoCliente):
    def __init__(self, adaptee: Adaptee):
        self.adaptee = adaptee

    def request(self):
        self.adaptee.especifica_request()


# 5 - UTILIZANDO O ADAPTER -----------------------------------
if __name__ == "__main__":
    adaptee = Adaptee()
    adapter = Adapter(adaptee)
    cliente = Cliente(adapter)

    # Fazendo uma requisição - Requisição Específica do Adaptee:
    cliente.cria_request()

File: TPII-2025-2/PadraoGoF/Estrutural/Adapter/Adapter01.java
// Exemplo de Adapter em Java
// O Adapter permite que interfaces incompatíveis trabalhem juntas.

// 1 - Interface do Cliente
interface AplicacaoCliente {
    void request();
}

// 2 - Serviço Existente
class Adaptee {
    void especificaRequest() {
        System.out.println("Requisição específica do Adaptee");
    }
}

// 3 - Adapter
class Adapter implements AplicacaoCliente {
    private Adaptee adaptee;
    Adapter(Adaptee adaptee) {
        this.adaptee = adaptee;
    }
    @Override
    public void request() {
        adaptee.especificaRequest();
    }
}

// 4 - Cliente que usa o Adapter
class Cliente {
    private AplicacaoCliente target;
    Cliente(AplicacaoCliente target) {
        this.target = target;
    }
    void criaRequest() {
        System.out.println("Cliente - Fazendo uma requisição");
        target.request();
    }
}

// 5 - Exemplo de uso
class MainAdapter {
    public static void main(String[] args) {
        Adaptee adaptee = new Adaptee();
        AplicacaoCliente adapter = new Adapter(adaptee);
        Cliente cliente = new Cliente(adapter);
        cliente.criaRequest();
    }
}
